backend/src/main/java/pl/edu/agh/niebieskiekotki/groups/GroupGenerator.java

```diff
@@ -0,0 +1,123 @@
+package pl.edu.agh.niebieskiekotki.groups;
+
+import org.jgrapht.alg.flow.DinicMFImpl;
+import org.jgrapht.graph.DefaultDirectedWeightedGraph;
+import org.jgrapht.graph.DefaultWeightedEdge;
+import org.springframework.stereotype.Component;
+import pl.edu.agh.niebieskiekotki.entitites.Student;
+import pl.edu.agh.niebieskiekotki.entitites.Term;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Component
+public class GroupGenerator {
+    public GenerationOutput generate(Map<Student, List<Term>> studentsTerms, int numGroups){
+        var network = new DefaultDirectedWeightedGraph<>(DefaultWeightedEdge.class);
+        var source = new Object();
+        var sink = new Object();
+
+        network.addVertex(source);
+        network.addVertex(sink);
+
+        Set<Student> allStudents = studentsTerms.keySet();
+
+        for (var student: allStudents){
+            network.addVertex(student);
+            network.addEdge(source, student);
+        }
+
+        Set<Term> allTerms = getAllTerms(studentsTerms);
+        Set<Term> candidateTerms = getCandidateTerms(studentsTerms, allTerms, numGroups);
+
+        for (var term: candidateTerms){
+            network.addVertex(term);
+        }
+
+        for (var student: allStudents){
+            for (var term: studentsTerms.get(student)){
+                if (candidateTerms.contains(term)) {
+                    network.addEdge(student, term);
+                }
+            }
+        }
+
+        double maxGroupSize = Math.round(((double) allStudents.size() / numGroups) + 0.5);
```

Comment:
Może warto dać większą możliwość na różne rozmiary w grupach? Nie jestem pewien jak dokładnie to by wpłynęło na algorytm, ale myślę że możemy dopuścić większą wariację w rozmiarach grup (jeżeli algorytm będzie w stanie to wykorzystać).